### PR TITLE
Support gemma model in pytorch engine

### PR DESCRIPTION
**Pull request by grimoire**
- Fix llama on transformers 4.38
- Support Gemma

- [x] #1075

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -94,6 +94,7 @@ For detailed inference benchmarks in more devices and more settings, please refe
 |      ChatGLM2      |    6B     |
 |       Falcon       | 7B - 180B |
 |         YI         | 6B - 34B  |
+|      Mistral       |    7B     |
```
Comment (zhyncs) [inline README.md:97]: Could we update the list of supported models for Gemma？

File: lmdeploy/cli/cli.py [hunk 2]
```diff
@@ -107,7 +107,7 @@ def list(args):
         if engine == 'pytorch':
             model_names = [
                 'llama', 'llama2', 'internlm', 'internlm2', 'baichuan2',
-                'chatglm2', 'falcon', 'yi'
+                'chatglm2', 'falcon', 'yi', 'mistral'
```
Comment (zhyncs) [inline lmdeploy/cli/cli.py:110]: `gemma` in `model_names`

**Comment by zhyncs** [comment]
I've tested it with `profile_restful_api.py` and `profile_throughput.py` for [google/gemma-2b-it](https://huggingface.co/google/gemma-2b-it) and it works well.

**Comment by zhyncs** [comment]
Hi @grimoire Have you compared the output results with those of the [transformers](https://github.com/huggingface/transformers/releases/tag/v4.38.1)?

File: README.md (hunk 1 above)
Comment (grimoire) [reply README.md:97]: Done

File: lmdeploy/cli/cli.py (hunk 2 above)
Comment (grimoire) [reply lmdeploy/cli/cli.py]: Done

**Comment by grimoire** [comment]
> Hi @grimoire Have you compared the output results with those of the [transformers](https://github.com/huggingface/transformers/releases/tag/v4.38.1)?

I have align the `GemmaDecodeLayer` with my [debug tools](https://github.com/InternLM/lmdeploy/blob/main/docs/en/advance/pytorch_new_model.md#debug-module) It can be matched without Patched RMS norm. Since the only different between `LlamaModel` and `GemmaModel` is `hidden_states = hidden_states * (self.config.hidden_size**0.5)`, I think that should be enough.

The patched RMS norm looks OK in chat API, I guess the mismatch of the norm is caused by the scaling of hidden_states.

**Comment by zhyncs** [comment]
Hi @grimoire Please resolve the conflicts.

**Comment by lvhan028** [comment]
UT failed
```
>           assert q.stride() == out_q.stride()
E           AttributeError: 'NoneType' object has no attribute 'stride'
```
Is it this PR's issue or our runner's issue?
@RunningLeon may check the runner.

**Comment by RunningLeon** [comment]
Tested Ok on with gemma and llma2
- chat cli
- pipeline
- api_server

**Review by RunningLeon — APPROVED** [review]
LGTM

**Comment by zhulinJulia24** [comment]
Error when running Baichuan2-7B-Chat model. It works in main branch.


script: 

```
from lmdeploy import pipeline, GenerationConfig, PytorchEngineConfig

backend_config = PytorchEngineConfig(session_len=2048)
gen_config = GenerationConfig(top_p=0.8,
                              top_k=40,
                              temperature=0.8,
                              max_new_tokens=1024)
pipe = pipeline('/nvme/qa_test_models/Baichuan2-7B-Chat',
                backend_config=backend_config)
prompts = [[{
    'role': 'user',
    'content': 'Hi, pls intro yourself'
}], [{
    'role': 'user',
    'content': 'Shanghai is'
}]]
response = pipe(prompts, gen_config=gen_config)
print(response)
```

error callstack.

```
Xformers is not installed correctly. If you want to use memory_efficient_attention to accelerate training use the following command to install Xformers
pip install xformers.
/home/zhulin1/miniconda3/envs/lmdeployv23/lib/python3.10/site-packages/torch/_utils.py:831: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
  return self.fget.__get__(instance, owner)()
Exception in thread Thread-1 (loop):
Traceback (most recent call last):
  File "/home/zhulin1/miniconda3/envs/lmdeployv23/lib/python3.10/threading.py", line 1009, in _bootstrap_inner
    self.run()
  File "/home/zhulin1/miniconda3/envs/lmdeployv23/lib/python3.10/threading.py", line 946, in run
    self._target(*self._args, **self._kwargs)
  File "/home/zhulin1/miniconda3/envs/lmdeployv23/lib/python3.10/site-packages/lmdeploy/pytorch/engine/engine.py", line 874, in loop
    step_tokens: Dict[int, InferOutput] = self.step(
  File "/home/zhulin1/miniconda3/envs/lmdeployv23/lib/python3.10/site-packages/lmdeploy/pytorch/engine/engine.py", line 668, in step
    output = self._model_forward(inputs,
  File "/home/zhulin1/miniconda3/envs/lmdeployv23/lib/python3.10/site-packages/lmdeploy/pytorch/engine/engine.py", line 641, in _model_forward
    return __forward(inputs)
  File "/home/zhulin1/miniconda3/envs/lmdeployv23/lib/python3.10/site-packages/lmdeploy/pytorch/engine/engine.py", line 583, in __forward
    return self.model_agent.forward(inputs,
  File "/home/zhulin1/miniconda3/envs/lmdeployv23/lib/python3.10/site-packages/lmdeploy/pytorch/engine/model_agent.py", line 522, in forward
    output = model_forward(
  File "/home/zhulin1/miniconda3/envs/lmdeployv23/lib/python3.10/site-packages/lmdeploy/pytorch/engine/model_agent.py", line 335, in model_forward
    output = patched_model.patched_forward(
  File "/home/zhulin1/miniconda3/envs/lmdeployv23/lib/python3.10/site-packages/lmdeploy/pytorch/models/patch.py", line 239, in __call__
    output = self._model(*args, **kwargs)
  File "/home/zhulin1/miniconda3/envs/lmdeployv23/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/zhulin1/miniconda3/envs/lmdeployv23/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/zhulin1/.cache/huggingface/modules/transformers_modules/Baichuan2-7B-Chat/modeling_baichuan.py", line 686, in forward
    outputs = self.model(
  File "/home/zhulin1/miniconda3/envs/lmdeployv23/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/zhulin1/miniconda3/envs/lmdeployv23/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/zhulin1/miniconda3/envs/lmdeployv23/lib/python3.10/site-packages/lmdeploy/pytorch/models/baichuan.py", line 470, in forward
    return self._continuous_batching_forward_7b(
  File "/home/zhulin1/miniconda3/envs/lmdeployv23/lib/python3.10/site-packages/lmdeploy/pytorch/models/baichuan.py", line 349, in _continuous_batching_forward_7b
    layer_outputs = decoder_layer(
  File "/home/zhulin1/miniconda3/envs/lmdeployv23/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/zhulin1/miniconda3/envs/lmdeployv23/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/zhulin1/.cache/huggingface/modules/transformers_modules/Baichuan2-7B-Chat/modeling_baichuan.py", line 273, in forward
    hidden_states, self_attn_weights, present_key_value = self.self_attn(
  File "/home/zhulin1/miniconda3/envs/lmdeployv23/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/zhulin1/miniconda3/envs/lmdeployv23/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/zhulin1/miniconda3/envs/lmdeployv23/lib/python3.10/site-packages/lmdeploy/pytorch/models/baichuan.py", line 87, in forward
    return self._contiguous_batching_forward(
  File "/home/zhulin1/miniconda3/envs/lmdeployv23/lib/python3.10/site-packages/lmdeploy/pytorch/models/baichuan.py", line 142, in _contiguous_batching_forward
    query_states, key_states, value_states = _rotary_emb_fn(
  File "/home/zhulin1/miniconda3/envs/lmdeployv23/lib/python3.10/site-packages/lmdeploy/pytorch/models/baichuan.py", line 131, in _rotary_emb_fn
    query_states, key_states = apply_rotary_pos_emb(
  File "/home/zhulin1/miniconda3/envs/lmdeployv23/lib/python3.10/site-packages/lmdeploy/pytorch/models/llama.py", line 43, in apply_rotary_pos_emb
    cos = cos.unsqueeze(unsqueeze_dim)
TypeError: unsqueeze(): argument 'dim' (position 1) must be int, not Tensor
2024-02-29 09:04:07,433 - lmdeploy - ERROR - Engine main loop stopped.
```

**Comment by grimoire** [comment]
> Gemma model need transformers>=4.38.0, Should we put a notice on readme?

https://github.com/InternLM/lmdeploy/blob/a6c25240814fbd79f98dea669f2b69e893161927/requirements/runtime.txt#L14 requirements has been updated

**Comment by grimoire** [comment]
@zhulinJulia24 
> Error when running Baichuan2-7B-Chat model. It works in main branch.

Fixed.

**Review by zhulinJulia24 — APPROVED** [review]
lgtm

cover:
    - llama-2-7b-chat
    - internlm-chat-7b
    - internlm-chat-20b
    - internlm2-chat-7b
    - internlm2-chat-20b
    - Baichuan2-7B-Chat
    - Baichuan2-13B-Chat
    - chatglm2-6b
    - falcon-7b
    - Yi-6B-Chat
    - internlm2-1_8b
    - internlm2-20b
    - Qwen1.5-7B-Chat
    - Mistral-7B-Instruct-v0.1
    - Mixtral-8x7B-Instruct-v0.1
    - gemma-7b-it
    - deepseek-moe-16b-chat
tools: hf chat, resful and pipeline basic case

**Pull request merged**